src/puppetlabs/trapperkeeper/internal.clj

```diff
@@ -380,20 +395,19 @@
    in reverse order (to account for dependency relationships)."
   [app-context :- (schema/atom a/TrapperkeeperAppContext)]
   (log/info "Beginning shutdown sequence")
-  (let [{:keys [ordered-services lifecycle-channel lifecycle-worker]} @app-context]
-    (doseq [[service-id s] (reverse ordered-services)]
-      (try
-        (run-lifecycle-fn! app-context s/stop "stop" service-id s)
-        (catch Exception e
-          (log/error e "Encountered error during shutdown sequence"))))
-    (when-not (async-prot/closed? lifecycle-worker)
-      (log/debug "Service shutdown complete, shutting down lifecycle worker")
-      (async/>!! lifecycle-channel {:type :shutdown
-                                    :task-function nil})
-      ;; wait for the channel to send us the return value so we know it's done
-      (async/<!! lifecycle-worker)
-      (log/debug "Lifecycle worker shutdown complete"))
-    (log/info "Finished shutdown sequence")))
+  (let [{:keys [ordered-services shutdown-channel lifecycle-worker]} @app-context
+        shutdown-fn (fn [] (doseq [[service-id s] (reverse ordered-services)]
+                             (try
+                               (run-lifecycle-fn! app-context s/stop "stop" service-id s)
+                               (catch Exception e
+                                 (log/error e "Encountered error during shutdown sequence")))))]
+    (log/trace "Putting shutdown message on shutdown channel.")
+    (async/>!! shutdown-channel {:type :shutdown
+                                 :task-function shutdown-fn})
+    ;; wait for the channel to send us the return value so we know it's done
+    (log/trace "Waiting for response to shutdown message from lifecycle worker.")
+    (if (not (nil? (async/<!! lifecycle-worker)))
+      (log/info "Finished shutdown sequence"))))
```

Comment:
Would it be helpful to still write some kind of message out if the shutdown is aborted - like for those trying to coordinate the number of "putting shutodnw message" / "waiting for response" messages to the corresponding result.  Maybe just a `(log/trace "Shutdown already requested, aborting additional shutdown attempt")`.  Open to whatever wording would sound best to you.